src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookSearch.ts

```diff
@@ -0,0 +1,693 @@
+	}
+
+	private createSearchResultIterator(collapseResults: ISearchConfigurationProperties['collapseResults']): Iterable<ITreeElement<RenderableMatch>> {
+		const folderMatches = this.searchResult.folderMatches()
+			.filter(fm => !fm.isEmpty())
+			.sort(searchMatchComparer);
+
+		if (folderMatches.length === 1) {
+			return this.createSearchFolderIterator(folderMatches[0], collapseResults);
+		}
+
+		return Iterable.map(folderMatches, folderMatch => {
+			const children = this.createSearchFolderIterator(folderMatch, collapseResults);
+			return <ITreeElement<RenderableMatch>>{ element: folderMatch, children };
+		});
+	}
+
+	private createSearchFolderIterator(folderMatch: FolderMatch, collapseResults: ISearchConfigurationProperties['collapseResults']): Iterable<ITreeElement<RenderableMatch>> {
+		const sortOrder = this.searchConfig.sortOrder;
+		const matches = folderMatch.matches().sort((a, b) => searchMatchComparer(a, b, sortOrder));
+
+		return Iterable.map(matches, fileMatch => {
+			//const children = this.createFileIterator(fileMatch);
+			let nodeExists = true;
+			try { this.tree.getNode(fileMatch); } catch (e) { nodeExists = false; }
+
+			const collapsed = nodeExists ? undefined :
+				(collapseResults === 'alwaysCollapse' || (fileMatch.matches().length > 10 && collapseResults !== 'alwaysExpand'));
+
+			return <ITreeElement<RenderableMatch>>{ element: fileMatch, undefined, collapsed, collapsible: false };
+		});
+	}
+
+	private createSearchFileIterator(fileMatch: FileMatch): Iterable<ITreeElement<RenderableMatch>> {
+		const matches = fileMatch.matches().sort(searchMatchComparer);
+		return Iterable.map(matches, r => (<ITreeElement<RenderableMatch>>{ element: r }));
+	}
+
+	private createSearchIterator(match: FolderMatch | FileMatch | SearchResult, collapseResults: ISearchConfigurationProperties['collapseResults']): Iterable<ITreeElement<RenderableMatch>> {
+		return match instanceof SearchResult ? this.createSearchResultIterator(collapseResults) :
+			match instanceof FolderMatch ? this.createSearchFolderIterator(match, collapseResults) :
+				this.createSearchFileIterator(match);
+	}
+
+	triggerSearchQueryChange(query: ITextQuery, excludePatternText: string, includePatternText: string, triggeredOnType: boolean, searchWidget: NotebookSearchWidget) {
+		this.viewModel.cancelSearch(true);
+
+		this.currentSearchQ = this.currentSearchQ
+			.then(() => this.startSearch(query, excludePatternText, includePatternText, triggeredOnType, searchWidget))
+			.then(() => undefined, () => undefined);
+	}
+
+	public saveState(): void {
+		const preserveCase = this.viewModel.preserveCase;
+		this.viewletState['query.preserveCase'] = preserveCase;
+
+		this.memento.saveMemento();
+
+		ViewPane.prototype.saveState.call(this);
+	}
+
+	dispose(): void {
+		this.isDisposed = true;
+		this.saveState();
+		this.treeSelectionChangeListener.dispose();
+		ViewPane.prototype.dispose.call(this);
+	}
+}
+
+class ToggleCollapseAndExpandAction extends Action {
```

Comment:
Just to check - these actions aren't something you can re-use from the vscode implementation? Or at least extend theirs and override certain behavior?

Reply:
I cannot reuse it but I can extend it. I'll do that in a separate PR.